src/rfactor/process.py

```diff
@@ -1,11 +1,110 @@
 import re
+from functools import wraps
 from pathlib import Path
 
 import numpy as np
 import pandas as pd
+from pandas import Timedelta
 from tqdm import tqdm
 
 
+def valid_column(rain, req_col):
+    """Input dataframe has valid required columns
+
+    Parameters
+    ----------
+    rain: pd.DataFrame
+        To test dataframe
+    req_col: set
+        Required columns in dataframe, e.g. {"datetime", "rain_mm"}
+    """
+    # test for columns
+    if not req_col.issubset(rain.columns):
+        raise KeyError(f"DataFrame should contain {req_col}  columns.")
+
+
+def valid_const_freq(rain):
+    """Check if rainfall inputdata has constant frequency
+
+    Parameters
+    ----------
+    rain: pandas.DataFrame
+    """
+    # constant frequency
+    if len(rain["datetime"].diff().unique()) > 2:
+        msg = (
+            "Timeseries resolution is not a strict constant, please define "
+            "rainfall timeseries with one temporal resolution."
+        )
+        raise IOError(msg)
+
+
+def valid_freq(freq, mode_test_freq):
+    """Valid frequency
+
+    Parameters
+    ----------
+    freq: pandas.DatetimeIndex.freq
+        Frequency
+    mode_test_freq: int
+        1: Frequency of input rainfall data must be at least minute resolution.
+        2: Frequency of input rainfall data must be 10-minute resolution.
```

Comment:
I think the `mode` approach is not intuitive. Can we simplify this? For example, remove the mode options and just require frequency >= 1 min and a user option to define if it needs a specific resolution? 

```
def valid_freq(df_freq, freq=None)
    ...
```
- `df_freq` always >=1min
- if `freq` not None, test if equal to this frequency

Reply:
Good idea, implemented as such